pkg/discovery/worker/metrics_collector.go

```diff
@@ -189,6 +190,55 @@ func collectContainersComputeResources(pod *v1.Pod) (float64, float64, float64,
 	return totalCPULimits, totalCPURequests, totalMemLimits, totalMemRequests
 }
 
+func (collector *MetricsCollector) CollectPodVolumeMetrics() []*repository.PodVolumeMetrics {
+	var podVolumeMetricsCollection []*repository.PodVolumeMetrics
+
+	var podToVolsMap map[string][]repository.MountedVolume
+	if collector.Cluster.PodToVolumesMap != nil {
+		podToVolsMap = collector.Cluster.PodToVolumesMap
+	}
+
+	metricsSink := collector.MetricsSink
+	//Iterate over all the pods in the collection
+	for _, pod := range collector.PodList {
+		podKey := fmt.Sprintf("%s/%s", pod.Namespace, pod.Name)
+		podVols, exists := podToVolsMap[podKey]
+		if !exists {
+			continue
+		}
+
+		for _, podVol := range podVols {
+			podVolumeMetric := repository.PodVolumeMetrics{}
+			volKey := util.PodVolumeMetricId(podKey, podVol.MountName)
+			found := false
+
+			metricUID := metrics.GenerateEntityResourceMetricUID(metrics.PodType, volKey,
+				metrics.StorageAmount, metrics.Used)
+			metric, _ := metricsSink.GetMetric(metricUID)
+			if metric != nil && metric.GetValue() != nil {
+				podVolumeMetric.Used = metric.GetValue().(float64)
+				found = true
+			}
+
+			metricUID = metrics.GenerateEntityResourceMetricUID(metrics.PodType, volKey,
+				metrics.StorageAmount, metrics.Capacity)
+			metric, _ = metricsSink.GetMetric(metricUID)
+			if metric != nil && metric.GetValue() != nil {
+				podVolumeMetric.Capacity = metric.GetValue().(float64)
+				found = true
+			}
+
+			if found == true {
```

Comment:
I wasn't (and still am not) aware of the requirement of the capacity being set. I am not sure what fails if a probe reports a commodity without capacity (and I think this can happen in kubeturbo too).
On the other hand I did write this code to handle cases where one or the other (for some reason) is not available. This is quite unlikely though, as the kubelet api, if it reports should ideally report both and thats the only point from where I might get one and not the other.

Reply:
I remember there was an issue where a commodity capacity is not set and an exception occurred inside market analysis like this one (https://vmturbo.atlassian.net/browse/OM-59514). The fix Endre eventually put there was to set a large default capacity. Would be good to set a default large capacity just in case, although ideally kubelet should be able to return both used and capacity.